File: forge/src/main/java/juuxel/adorn/platform/forge/registrar/ForgeRegistrarFactory.java
package juuxel.adorn.platform.forge.registrar;

import juuxel.adorn.lib.registry.KeyedRegistrar;
import juuxel.adorn.lib.registry.Registrar;
import juuxel.adorn.lib.registry.RegistrarFactory;
import net.minecraft.network.syncher.EntityDataSerializer;
import net.minecraft.core.Registry;
import net.minecraft.resources.ResourceKey;
import net.neoforged.neoforge.registries.NeoForgeRegistries;

public final class ForgeRegistrarFactory implements RegistrarFactory {
    @Override
    public <T> KeyedRegistrar<T> create(ResourceKey<Registry<T>> registryKey) {
        return new DeferredRegistrar<>(registryKey);
    }

    @Override
    public Registrar<EntityDataSerializer<?>> createForTrackedDataHandlers() {
        return create(NeoForgeRegistries.Keys.ENTITY_DATA_SERIALIZERS);
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/registrar/ForgeRegistrarFactory.kt
package juuxel.adorn.platform.forge.registrar

import juuxel.adorn.lib.registry.Registrar
import juuxel.adorn.lib.registry.RegistrarFactory
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey

class ForgeRegistrarFactory : RegistrarFactory {
    override fun <T> create(registryKey: RegistryKey<Registry<T>>): Registrar<T> =
        DeferredRegistrar(registryKey)
}
